onadata/apps/api/viewsets/data_viewset.py

```diff
@@ -575,18 +576,30 @@ def enketo(self, request, *args, **kwargs):
         if isinstance(self.object, XForm):
             raise ParseError(_("Data id not provided."))
         elif isinstance(self.object, Instance):
-            if request.user.has_perm("change_xform", self.object.xform):
-                return_url = request.query_params.get('return_url')
+            return_url = request.query_params.get('return_url')
+            action = request.query_params.get('action', 'edit')
+            if action == 'edit' and request.user.has_perm(
+                CAN_CHANGE_XFORM, self.object.xform
+            ):
                 if not return_url:
                     raise ParseError(_("return_url not provided."))
-
-                try:
-                    data["url"] = get_enketo_edit_url(
-                        request, self.object, return_url)
-                except EnketoError as e:
-                    data['detail'] = "{}".format(e)
+            elif action == 'view':
+                # Allow anonymous access to view-only if the data has been made
+                # public
+                if (
+                    request.user.is_anonymous
+                    and not self.object.xform.shared_data
+                ):
+                    raise Http404
```

Comment:
For the record, as per our discussion, this code will be refactored to move the permission verification within the permission class.